File: tests/test_evaluator.py
"""Unit tests for the core SC-1/NC-1/NC-2 evaluator."""

from __future__ import annotations

from blindspotcheck.evaluator import (
    ArchitectureReport,
    Verdict,
    evaluate_architecture,
)

# Minimal two-owner architecture template used throughout tests.
BASE_ARCH = {
    "meta": {"schema_version": "1", "domain": "test"},
    "asset_owners": [
        {"id": "APT", "role": "AO"},
        {"id": "ALN", "role": "AO"},
        {"id": "VND", "role": "AO"},
    ],
}


def _arch(**overrides) -> dict:
    base = {**BASE_ARCH}
    base.update(overrides)
    base.setdefault("sp_relations", [])
    base.setdefault("zone_authorities", [])
    return base


def test_blind_spot_three_conditions_hold() -> None:
    """SC-1 AND NC-1 AND NC-2: classification must be BLIND_SPOT."""
    arch = _arch(
        conduits=[
            {
                "id": "CD-X",
                "from": {"owner": "APT", "zone": "Z-APT"},
                "to": {"owner": "ALN", "zone": "Z-ALN"},
            }
        ],
        zone_authorities=[
            {"zone": "Z-APT", "org": "APT"},
            {"zone": "Z-ALN", "org": "ALN"},
        ],
    )
    report = evaluate_architecture(arch)
    (r,) = report.results
    assert r.sc1 is True
    assert r.nc1 is True
    assert r.nc2 is True
    assert r.verdict is Verdict.BLIND_SPOT
    assert r.mitigation, "blind spots must expose mitigation options"


def test_resolved_by_sp_when_relation_covers() -> None:
    """Cross-AO conduit with SP-AO relation => RESOLVED_BY_SP."""
    arch = _arch(
        conduits=[
            {
                "id": "CD-01",
                "from": {"owner": "VND", "zone": "Z-VND"},
                "to": {"owner": "APT", "zone": "Z-APT"},
            }
        ],
        sp_relations=[{"sp": "VND", "ao": "APT", "scope": ["CD-01"]}],
        zone_authorities=[
            {"zone": "Z-VND", "org": "VND"},
            {"zone": "Z-APT", "org": "APT"},
        ],
    )
    (r,) = evaluate_architecture(arch).results
    assert r.sc1 is True
    assert r.nc1 is False  # covered by SP-AO relationship
    assert r.verdict is Verdict.RESOLVED_BY_SP


def test_sp_relation_unscoped_covers_all_matching_pairs() -> None:
    """sp_relations without `scope` must cover all conduits with that pair."""
    arch = _arch(
        conduits=[
            {
                "id": "CD-10",
                "from": {"owner": "VND", "zone": "Z-VND"},
                "to": {"owner": "APT", "zone": "Z-APT"},
            },
            {
                "id": "CD-11",
                "from": {"owner": "VND", "zone": "Z-VND"},
                "to": {"owner": "APT", "zone": "Z-APT"},
            },
        ],
        sp_relations=[{"sp": "VND", "ao": "APT"}],
        zone_authorities=[
            {"zone": "Z-VND", "org": "VND"},
            {"zone": "Z-APT", "org": "APT"},
        ],
    )
    results = {r.conduit_id: r for r in evaluate_architecture(arch).results}
    assert results["CD-10"].verdict is Verdict.RESOLVED_BY_SP
    assert results["CD-11"].verdict is Verdict.RESOLVED_BY_SP


def test_no_cross_ao_when_same_owner_endpoints() -> None:
    """Same-owner endpoints => SC-1 fails => NO_CROSS_AO."""
    arch = _arch(
        conduits=[
            {
                "id": "CD-INTRA",
                "from": {"owner": "APT", "zone": "Z-APT"},
                "to": {"owner": "APT", "zone": "Z-APT"},
            }
        ],
        zone_authorities=[{"zone": "Z-APT", "org": "APT"}],
    )
    (r,) = evaluate_architecture(arch).results
    assert r.sc1 is False
    assert r.verdict is Verdict.NO_CROSS_AO


def test_borderline_when_same_zone_org_bridges() -> None:
    """SC-1 AND NC-1 true, but a single org defines both zones => NC-2 fails => BORDERLINE."""
    # Contrived: APT and ALN are distinct AOs, but ALN's endpoint sits in a
    # zone designated by APT (shared tenant VLAN).
    arch = _arch(
        conduits=[
            {
                "id": "CD-BORD",
                "from": {"owner": "APT", "zone": "Z-APT"},
                "to": {"owner": "ALN", "zone": "Z-APT-SHARED"},
            }
        ],
        zone_authorities=[
            {"zone": "Z-APT", "org": "APT"},
            {"zone": "Z-APT-SHARED", "org": "APT"},
        ],
    )
    (r,) = evaluate_architecture(arch).results
    assert r.sc1 is True
    assert r.nc1 is True
    assert r.nc2 is False
    assert r.verdict is Verdict.BORDERLINE


def test_nc2_fallback_when_no_zone_metadata() -> None:
    """When zones are unspecified, NC-2 falls back to SC-1 parity."""
    arch = _arch(
        conduits=[
            {
                "id": "CD-NOZONE",
                "from": {"owner": "APT"},
                "to": {"owner": "ALN"},
            }
        ],
    )
    (r,) = evaluate_architecture(arch).results
    assert r.sc1 is True
    assert r.nc2 is True  # fallback: distinct owners => distinct partitioning


def test_report_distribution_sums_to_total() -> None:
    """Distribution counts must equal total number of conduits."""
    arch = _arch(
        conduits=[
            {"id": "C1", "from": {"owner": "APT"}, "to": {"owner": "ALN"}},
            {"id": "C2", "from": {"owner": "APT"}, "to": {"owner": "APT"}},
            {"id": "C3", "from": {"owner": "VND"}, "to": {"owner": "APT"}},
        ],
        sp_relations=[{"sp": "VND", "ao": "APT", "scope": ["C3"]}],
    )
    report = evaluate_architecture(arch)
    dist = report.distribution()
    assert sum(dist.values()) == len(report.results) == 3


def test_empty_conduits_list_allowed() -> None:
    """Zero conduits should produce an empty report, not an error."""
    # Schema requires minItems:1, but evaluator itself must not crash on empty.
    arch = _arch(conduits=[])
    report = evaluate_architecture(arch)
    assert isinstance(report, ArchitectureReport)
    assert report.results == ()
    assert sum(report.distribution().values()) == 0


# ---------------------------------------------------------------------------
# sp_subtype (Batch 4): integration-only SP-AO does NOT satisfy NC-1
# ---------------------------------------------------------------------------


def test_sp_relation_defaults_to_both_subtype_backward_compat() -> None:
    """
    A relation without sp_subtype must behave as the pre-Batch-4 tool did:
    cover the conduit for NC-1 purposes (verdict = resolved-by-sp).
    """
    arch = _arch(
        conduits=[{
            "id": "CD-X",
            "from": {"owner": "VND", "zone": "Z-VND"},
            "to":   {"owner": "APT", "zone": "Z-APT"},
        }],
        sp_relations=[{"sp": "VND", "ao": "APT", "scope": ["CD-X"]}],
        zone_authorities=[
            {"zone": "Z-VND", "org": "VND"},
            {"zone": "Z-APT", "org": "APT"},
        ],
    )
    (r,) = evaluate_architecture(arch).results
    assert r.nc1 is False, "default subtype 'both' must cover (NC-1 false = covered)"
    assert r.verdict is Verdict.RESOLVED_BY_SP


def test_maintenance_sp_satisfies_nc1() -> None:
    """Maintenance SP (62443-2-4 Cl. 3.1.13 + SP.08.02 BR) covers monitoring."""
    arch = _arch(
        conduits=[{
            "id": "CD-X",
            "from": {"owner": "VND", "zone": "Z-VND"},
            "to":   {"owner": "APT", "zone": "Z-APT"},
        }],
        sp_relations=[{
            "sp": "VND", "ao": "APT", "scope": ["CD-X"],
            "sp_subtype": "maintenance",
        }],
        zone_authorities=[
            {"zone": "Z-VND", "org": "VND"},
            {"zone": "Z-APT", "org": "APT"},
        ],
    )
    (r,) = evaluate_architecture(arch).results
    assert r.verdict is Verdict.RESOLVED_BY_SP


def test_integration_only_sp_does_not_satisfy_nc1() -> None:
    """
    Integration-only SP (62443-2-4 Cl. 3.1.12) covers design/install/
    commissioning, NOT monitoring. Paper §2.3: "the Clause 3.1.13 definition
    explicitly includes 'security monitoring' as a maintenance SP activity."
    The conduit should therefore still be classified as a structural blind
    spot when SC-1 and NC-2 also hold.
    """
    arch = _arch(
        conduits=[{
            "id": "CD-X",
            "from": {"owner": "VND", "zone": "Z-VND"},
            "to":   {"owner": "APT", "zone": "Z-APT"},
        }],
        sp_relations=[{
            "sp": "VND", "ao": "APT", "scope": ["CD-X"],
            "sp_subtype": "integration",
        }],
        zone_authorities=[
            {"zone": "Z-VND", "org": "VND"},
            {"zone": "Z-APT", "org": "APT"},
        ],
    )
    (r,) = evaluate_architecture(arch).results
    assert r.sc1 is True
    assert r.nc1 is True, (
        "integration-only SP-AO must leave NC-1 satisfied "
        "(integration SPs do not reach monitoring per Cl. 3.1.12)"
    )
    assert r.nc2 is True
    assert r.verdict is Verdict.BLIND_SPOT


def test_nc2_zone_half_declared_falls_back_conservative() -> None:
    """
    When one endpoint carries a zone reference but the other doesn't,
    evaluator.py uses the "conservative" NC-2 = True branch: it treats
    the conduit as partitioned by distinct authorities because we have
    no basis to assert a single authority bridges both sides.
    """
    arch = _arch(
        conduits=[{
            "id": "CD-HALF",
            "from": {"owner": "VND", "zone": "Z-VND"},
            "to":   {"owner": "APT"},  # no zone on target
        }],
        zone_authorities=[
            {"zone": "Z-VND", "org": "VND"},
        ],
    )
    (r,) = evaluate_architecture(arch).results
    # The branch fires because fromZone and toZone aren't both truthy,
    # so evaluator falls through to the "no zone metadata" fallback
    # (nc2 = sc1 = True).
    assert r.sc1 is True
    assert r.nc2 is True


def test_nc2_zone_declared_but_org_missing_is_conservative() -> None:
    """
    Both endpoints carry zones but one of those zones is NOT in
    zone_authorities. The evaluator then has no org for that side and
    conservatively declares NC-2 true.
    """
    arch = _arch(
        conduits=[{
            "id": "CD-ORPHAN-ZONE",
            "from": {"owner": "VND", "zone": "Z-VND"},
            "to":   {"owner": "APT", "zone": "Z-UNKNOWN"},
        }],
        zone_authorities=[
            {"zone": "Z-VND", "org": "VND"},
            # Z-UNKNOWN intentionally missing -> schema allows it because
            # the referential-integrity check only forbids zones in
            # conduits that aren't in zone_authorities when both sides
            # attempt to resolve - except in this tooling we'd normally
            # catch it; here we construct via direct evaluator call to
            # stress the fallback path.
        ],
    )
    # Architecture schema validation would normally reject this, but
    # the evaluator itself must still handle the edge. Feed directly.
    (r,) = evaluate_architecture(arch).results
    assert r.nc2 is True  # missing org -> conservative True


def test_self_loop_is_no_cross_ao() -> None:
    """Edge source == target is always single-AO. Evaluator must not crash."""
    arch = _arch(
        conduits=[{
            "id": "CD-SELF",
            "from": {"owner": "APT"},
            "to":   {"owner": "APT"},
        }],
    )
    (r,) = evaluate_architecture(arch).results
    assert r.sc1 is False
    assert r.verdict is Verdict.NO_CROSS_AO


def test_coexisting_integration_and_maintenance_relations_resolve() -> None:
    """
    If the same (sp, ao) pair has both an integration relation AND a
    maintenance relation, the maintenance side alone is enough to cover
    the conduit for monitoring purposes.
    """
    arch = _arch(
        conduits=[{
            "id": "CD-X",
            "from": {"owner": "VND", "zone": "Z-VND"},
            "to":   {"owner": "APT", "zone": "Z-APT"},
        }],
        sp_relations=[
            {"sp": "VND", "ao": "APT", "scope": ["CD-X"], "sp_subtype": "integration"},
            {"sp": "VND", "ao": "APT", "scope": ["CD-X"], "sp_subtype": "maintenance"},
        ],
        zone_authorities=[
            {"zone": "Z-VND", "org": "VND"},
            {"zone": "Z-APT", "org": "APT"},
        ],
    )
    (r,) = evaluate_architecture(arch).results
    assert r.verdict is Verdict.RESOLVED_BY_SP

File: tests/web/test_evaluator.mjs
// Unit tests for the JS evaluator + schema validator.
// Uses only the Node built-in test runner (node:test) and node:assert.
//
// Run with:   node --test tests/web/

import { test } from 'node:test';
import assert from 'node:assert/strict';

import { evaluateArchitecture, Verdict } from '../../web/evaluator.js';
import {
  validateArchitecture,
  SchemaError,
} from '../../web/schema-validator.js';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

const BASE_ARCH = Object.freeze({
  meta: { schema_version: '1', domain: 'test' },
  asset_owners: [
    { id: 'APT', role: 'AO' },
    { id: 'ALN', role: 'AO' },
    { id: 'VND', role: 'AO' },
  ],
});

function arch(overrides) {
  const base = { ...BASE_ARCH };
  Object.assign(base, overrides);
  if (!('sp_relations' in base)) base.sp_relations = [];
  if (!('zone_authorities' in base)) base.zone_authorities = [];
  return base;
}

// ---------------------------------------------------------------------------
// Evaluator: classification logic
// ---------------------------------------------------------------------------

test('blind-spot when SC-1, NC-1, NC-2 all hold', () => {
  const a = arch({
    conduits: [
      {
        id: 'CD-X',
        from: { owner: 'APT', zone: 'Z-APT' },
        to: { owner: 'ALN', zone: 'Z-ALN' },
      },
    ],
    zone_authorities: [
      { zone: 'Z-APT', org: 'APT' },
      { zone: 'Z-ALN', org: 'ALN' },
    ],
  });
  const [r] = evaluateArchitecture(a).results;
  assert.equal(r.sc1, true);
  assert.equal(r.nc1, true);
  assert.equal(r.nc2, true);
  assert.equal(r.verdict, Verdict.BLIND_SPOT);
  assert.ok(r.mitigation.length > 0, 'blind spots must expose mitigation options');
});

test('resolved-by-sp when a covering SP-AO relation exists', () => {
  const a = arch({
    conduits: [
      {
        id: 'CD-01',
        from: { owner: 'VND', zone: 'Z-VND' },
        to: { owner: 'APT', zone: 'Z-APT' },
      },
    ],
    sp_relations: [{ sp: 'VND', ao: 'APT', scope: ['CD-01'] }],
    zone_authorities: [
      { zone: 'Z-VND', org: 'VND' },
      { zone: 'Z-APT', org: 'APT' },
    ],
  });
  const [r] = evaluateArchitecture(a).results;
  assert.equal(r.sc1, true);
  assert.equal(r.nc1, false);
  assert.equal(r.verdict, Verdict.RESOLVED_BY_SP);
});

test('sp_relations without scope covers all matching pairs', () => {
  const a = arch({
    conduits: [
      {
        id: 'CD-10',
        from: { owner: 'VND', zone: 'Z-VND' },
        to: { owner: 'APT', zone: 'Z-APT' },
      },
      {
        id: 'CD-11',
        from: { owner: 'VND', zone: 'Z-VND' },
        to: { owner: 'APT', zone: 'Z-APT' },
      },
    ],
    sp_relations: [{ sp: 'VND', ao: 'APT' }],
    zone_authorities: [
      { zone: 'Z-VND', org: 'VND' },
      { zone: 'Z-APT', org: 'APT' },
    ],
  });
  const byId = Object.fromEntries(
    evaluateArchitecture(a).results.map((r) => [r.conduit_id, r]),
  );
  assert.equal(byId['CD-10'].verdict, Verdict.RESOLVED_BY_SP);
  assert.equal(byId['CD-11'].verdict, Verdict.RESOLVED_BY_SP);
});

test('no-cross-ao when endpoints share the same owner', () => {
  const a = arch({
    conduits: [
      {
        id: 'CD-INTRA',
        from: { owner: 'APT', zone: 'Z-APT' },
        to: { owner: 'APT', zone: 'Z-APT' },
      },
    ],
    zone_authorities: [{ zone: 'Z-APT', org: 'APT' }],
  });
  const [r] = evaluateArchitecture(a).results;
  assert.equal(r.sc1, false);
  assert.equal(r.verdict, Verdict.NO_CROSS_AO);
});

test('borderline when a single org bridges both endpoint zones', () => {
  const a = arch({
    conduits: [
      {
        id: 'CD-BORD',
        from: { owner: 'APT', zone: 'Z-APT' },
        to: { owner: 'ALN', zone: 'Z-APT-SHARED' },
      },
    ],
    zone_authorities: [
      { zone: 'Z-APT', org: 'APT' },
      { zone: 'Z-APT-SHARED', org: 'APT' },
    ],
  });
  const [r] = evaluateArchitecture(a).results;
  assert.equal(r.sc1, true);
  assert.equal(r.nc1, true);
  assert.equal(r.nc2, false);
  assert.equal(r.verdict, Verdict.BORDERLINE);
});

test('NC-2 falls back to SC-1 when zone metadata is absent', () => {
  const a = arch({
    conduits: [
      {
        id: 'CD-NOZONE',
        from: { owner: 'APT' },
        to: { owner: 'ALN' },
      },
    ],
  });
  const [r] = evaluateArchitecture(a).results;
  assert.equal(r.sc1, true);
  assert.equal(r.nc2, true);
});

test('distribution counts sum to the total number of conduits', () => {
  const a = arch({
    conduits: [
      { id: 'C1', from: { owner: 'APT' }, to: { owner: 'ALN' } },
      { id: 'C2', from: { owner: 'APT' }, to: { owner: 'APT' } },
      { id: 'C3', from: { owner: 'VND' }, to: { owner: 'APT' } },
    ],
    sp_relations: [{ sp: 'VND', ao: 'APT', scope: ['C3'] }],
  });
  const report = evaluateArchitecture(a);
  const dist = report.distribution();
  const total = Object.values(dist).reduce((s, v) => s + v, 0);
  assert.equal(total, report.results.length);
  assert.equal(total, 3);
});

test('default sp_subtype "both" covers the conduit (backward compat)', () => {
  const a = arch({
    conduits: [
      { id: 'CD-X',
        from: { owner: 'VND', zone: 'Z-VND' },
        to:   { owner: 'APT', zone: 'Z-APT' } },
    ],
    sp_relations: [{ sp: 'VND', ao: 'APT', scope: ['CD-X'] }],
    zone_authorities: [
      { zone: 'Z-VND', org: 'VND' },
      { zone: 'Z-APT', org: 'APT' },
    ],
  });
  const [r] = evaluateArchitecture(a).results;
  assert.equal(r.nc1, false);
  assert.equal(r.verdict, Verdict.RESOLVED_BY_SP);
});

test('maintenance sp_subtype resolves NC-1', () => {
  const a = arch({
    conduits: [
      { id: 'CD-X',
        from: { owner: 'VND', zone: 'Z-VND' },
        to:   { owner: 'APT', zone: 'Z-APT' } },
    ],
    sp_relations: [{ sp: 'VND', ao: 'APT', scope: ['CD-X'], sp_subtype: 'maintenance' }],
    zone_authorities: [
      { zone: 'Z-VND', org: 'VND' },
      { zone: 'Z-APT', org: 'APT' },
    ],
  });
  const [r] = evaluateArchitecture(a).results;
  assert.equal(r.verdict, Verdict.RESOLVED_BY_SP);
});

test('integration-only sp_subtype does NOT resolve NC-1 (paper §2.3)', () => {
  // Integration SP (62443-2-4 Cl. 3.1.12) covers design/install/commissioning
  // only; SP.08.02 BR monitoring obligation attaches to maintenance (Cl.
  // 3.1.13). The conduit must still classify as a structural blind spot.
  const a = arch({
    conduits: [
      { id: 'CD-X',
        from: { owner: 'VND', zone: 'Z-VND' },
        to:   { owner: 'APT', zone: 'Z-APT' } },
    ],
    sp_relations: [{ sp: 'VND', ao: 'APT', scope: ['CD-X'], sp_subtype: 'integration' }],
    zone_authorities: [
      { zone: 'Z-VND', org: 'VND' },
      { zone: 'Z-APT', org: 'APT' },
    ],
  });
  const [r] = evaluateArchitecture(a).results;
  assert.equal(r.sc1, true);
  assert.equal(r.nc1, true, 'integration-only SP-AO must leave NC-1 satisfied');
  assert.equal(r.nc2, true);
  assert.equal(r.verdict, Verdict.BLIND_SPOT);
});

test('coexisting integration + maintenance relations resolve', () => {
  const a = arch({
    conduits: [
      { id: 'CD-X',
        from: { owner: 'VND', zone: 'Z-VND' },
        to:   { owner: 'APT', zone: 'Z-APT' } },
    ],
    sp_relations: [
      { sp: 'VND', ao: 'APT', scope: ['CD-X'], sp_subtype: 'integration' },
      { sp: 'VND', ao: 'APT', scope: ['CD-X'], sp_subtype: 'maintenance' },
    ],
    zone_authorities: [
      { zone: 'Z-VND', org: 'VND' },
      { zone: 'Z-APT', org: 'APT' },
    ],
  });
  const [r] = evaluateArchitecture(a).results;
  assert.equal(r.verdict, Verdict.RESOLVED_BY_SP);
});

test('non-AO role (integrator) does not produce a blind spot', () => {
  const a = {
    meta: { schema_version: '1', domain: 't' },
    asset_owners: [
      { id: 'APT', role: 'AO' },
      { id: 'INT', role: 'integrator' },
    ],
    conduits: [
      { id: 'CD-INT', from: { owner: 'APT' }, to: { owner: 'INT' } },
    ],
    sp_relations: [],
    zone_authorities: [],
  };
  const [r] = evaluateArchitecture(a).results;
  assert.equal(r.nc1, false);
  assert.notEqual(r.verdict, Verdict.BLIND_SPOT);
});

// ---------------------------------------------------------------------------
// Schema validator: accepts minimal valid, rejects malformed
// ---------------------------------------------------------------------------

test('validator accepts the minimal valid architecture', () => {
  assert.doesNotThrow(() => validateArchitecture(arch({
    conduits: [{ id: 'C1', from: { owner: 'APT' }, to: { owner: 'ALN' } }],
  })));
});

test('validator rejects unknown schema_version', () => {
  const bad = arch({
    meta: { schema_version: '99', domain: 'test' },
    conduits: [{ id: 'C1', from: { owner: 'APT' }, to: { owner: 'ALN' } }],
  });
  assert.throws(() => validateArchitecture(bad), SchemaError);
});

test('validator rejects additionalProperties', () => {
  const bad = arch({
    conduits: [
      {
        id: 'C1',
        from: { owner: 'APT' },
        to: { owner: 'ALN' },
        typo_field: 'should fail',
      },
    ],
  });
  assert.throws(() => validateArchitecture(bad), SchemaError);
});

test('validator rejects bad role enum', () => {
  const bad = arch({
    asset_owners: [
      { id: 'APT', role: 'HACKER' },
      { id: 'ALN', role: 'AO' },
    ],
    conduits: [{ id: 'C1', from: { owner: 'APT' }, to: { owner: 'ALN' } }],
  });
  assert.throws(() => validateArchitecture(bad), SchemaError);
});

test('validator rejects bad id pattern (leading digit)', () => {
  const bad = arch({
    conduits: [
      {
        id: '1Cstart-with-digit',
        from: { owner: 'APT' },
        to: { owner: 'ALN' },
      },
    ],
  });
  assert.throws(() => validateArchitecture(bad), SchemaError);
});

test('validator rejects endpoint owner not in asset_owners', () => {
  const bad = arch({
    conduits: [{ id: 'C1', from: { owner: 'NOPE' }, to: { owner: 'ALN' } }],
  });
  assert.throws(() => validateArchitecture(bad), SchemaError);
});

test('validator rejects duplicate conduit ids', () => {
  const bad = arch({
    conduits: [
      { id: 'DUP', from: { owner: 'APT' }, to: { owner: 'ALN' } },
      { id: 'DUP', from: { owner: 'ALN' }, to: { owner: 'APT' } },
    ],
  });
  assert.throws(() => validateArchitecture(bad), SchemaError);
});

test('validator rejects sp_relations.scope referencing unknown conduit', () => {
  const bad = arch({
    conduits: [{ id: 'C1', from: { owner: 'APT' }, to: { owner: 'ALN' } }],
    sp_relations: [{ sp: 'APT', ao: 'ALN', scope: ['CD-UNKNOWN'] }],
  });
  assert.throws(() => validateArchitecture(bad), SchemaError);
});

// ---------------------------------------------------------------------------
// Fail-closed on unknown schema keywords (Stage G / Batch 1)
// ---------------------------------------------------------------------------

test('validator throws when a schema uses an unsupported Draft 2020-12 keyword', async () => {
  // Construct a synthetic schema that uses `oneOf`, a keyword the web
  // validator deliberately does NOT implement. A silent-skip would let
  // invalid data pass on the browser side while the Python jsonschema
  // validator rejected it; we want loud failure instead.
  const schema = {
    type: 'object',
    required: ['meta', 'asset_owners', 'conduits'],
    additionalProperties: false,
    properties: {
      meta: {
        type: 'object',
        properties: {
          schema_version: { type: 'string', oneOf: [{ const: '1' }] },
          domain: { type: 'string' },
        },
        required: ['schema_version', 'domain'],
      },
      asset_owners: {
        type: 'array',
        items: { type: 'object', properties: { id: { type: 'string' }, role: { type: 'string' } } },
      },
      conduits: { type: 'array', items: { type: 'object' } },
    },
  };
  const data = { meta: { schema_version: '1', domain: 't' }, asset_owners: [], conduits: [] };
  assert.throws(
    () => validateArchitecture(data, schema),
    /Unsupported JSON Schema keyword 'oneOf'/,
  );
});

test('the bundled ARCHITECTURE_SCHEMA_V1 uses only keywords the validator knows', async () => {
  // Meta-sweep: walk the bundled schema and assert every key is in
  // KNOWN_SCHEMA_KEYWORDS. If someone adds a new constraint keyword to
  // schema.js without extending the validator, this test fires.
  const { ARCHITECTURE_SCHEMA_V1 } = await import('../../web/schema.js');
  const { KNOWN_SCHEMA_KEYWORDS } = await import('../../web/schema-validator.js');

  const unknown = [];
  const walk = (node, where) => {
    if (!node || typeof node !== 'object' || Array.isArray(node)) return;
    for (const key of Object.keys(node)) {
      if (!KNOWN_SCHEMA_KEYWORDS.has(key)) {
        unknown.push(`${where}.${key}`);
      }
    }
    // Recurse into sub-schemas the validator will walk.
    if (node.properties) {
      for (const [k, v] of Object.entries(node.properties)) walk(v, `${where}.properties.${k}`);
    }
    if (node.items) walk(node.items, `${where}.items`);
    if (node.$defs) {
      for (const [k, v] of Object.entries(node.$defs)) walk(v, `${where}.$defs.${k}`);
    }
  };
  walk(ARCHITECTURE_SCHEMA_V1, '');
  assert.deepEqual(
    unknown,
    [],
    `ARCHITECTURE_SCHEMA_V1 uses keywords the validator does not handle: ${unknown.join(', ')}`,
  );
});
